### PR TITLE
Service connection support for managed identity and federated workload

### PR DESCRIPTION
**Pull request by jaredfholgate**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

This PR adds functionality to enable users to create service connections that support the Managed Identity and Workload Federation schemes. 

In summary it adds a new `service_endpoint_authentication_scheme` argument that accepts ones of 3 values:

- ServicePrincipal
- ManagedServiceIdentity
- WorkloadIdentityFederation

Based on this it then generates the relevant type of service connection using the supplied parameters.

There are essentially these 12 types of azurerm Service Connection available following this change:

- ServicePrincipal::Manual::ManagementGroupScope
- ServicePrincipal::Manual::SubscriptionScope
- ServicePrincipal::Automatic::ManagementGroupScope
- ServicePrincipal::Automatic::SubscriptionScope
- ServicePrincipal::Automatic::SubscriptionScopeWithResourceGroupPerms
- WorkloadIdentityFederation::Manual::ManagementGroupScope
- WorkloadIdentityFederation::Manual::SubscriptionScope
- WorkloadIdentityFederation::Automatic::ManagementGroupScope
- WorkloadIdentityFederation::Automatic::SubscriptionScope
- WorkloadIdentityFederation::Automatic::SubscriptionScopeWithResourceGroupPerms
- ManagedServiceIdentity::ManagementGroupScope
- ManagedServiceIdentity::SubscriptionScope

Issue Number: #101 

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Any relevant logs, error output, etc?

(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

## Other information

Here is evidence of the acceptance tests passing:
image

image

### CODE REVIEW

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go:103]: Does this means the project is not necessary or the project name can be any dummy values?

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go]: this will be breaking change to the existing resources, need add a state migration to keep the compatible
https://github.com/microsoft/terraform-provider-azuredevops/blob/main/azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go#L71

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go
Comment (kboom) [inline azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go]: Why are these lines commented?

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go
Comment (kboom) [inline azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go]: The type would be `AzureRM` and what `ServicePrincipal` means here is `Service Endpoint Authentication Scheme`. Applies to the rest of the usages as well.

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go
Comment (kboom) [inline azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go:148]: There is also an option to create a draft of a service connection with this type, in case someone wants to configure the federation entry in the service principal (or MI) only **after** they reserve the name. There is some merit to it, since if they don't, then anyone else with administrative rights in the projects could "hijack" that name and could gain access to the AAD object. Since this is all automation it probably doesn't make sense to offer this since we probably create an unverified endpoint first and only then create the federation for it, but just flagging this up.

File: azuredevops/internal/acceptancetests/testutils/hcl.go
Comment (kboom) [inline azuredevops/internal/acceptancetests/testutils/hcl.go]: nit: formatting

File: azuredevops/internal/acceptancetests/testutils/hcl.go
Comment (kboom) [inline azuredevops/internal/acceptancetests/testutils/hcl.go:475]: nit: formatting

File: azuredevops/internal/acceptancetests/testutils/hcl.go
Comment (kboom) [inline azuredevops/internal/acceptancetests/testutils/hcl.go]: nit: formatting

File: azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go
Comment (kboom) [inline azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go]: nit: It's not a type in ADO semantics but an auth scheme.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go
Comment (kboom) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go:75]: Just heads up - we will be deprecating this sometime in the future. WIF will be the standard recommended option.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go
Comment (kboom) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go:262]: There is no `serviceprincipalkey` for WIF. Why are we setting it? Then if someone has set it up, ADO would reject it.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm_test.go
Comment (kboom) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm_test.go:185]: I'd recommend not setting it at all. For a time being it will continue failing, but once our fix is rolled out it will work without issues.

File: website/docs/r/serviceendpoint_azurerm.html.markdown
Comment (kboom) [inline website/docs/r/serviceendpoint_azurerm.html.markdown]: Not only not required, but not permitted (server would reject such calls).

**Review by kboom — APPROVED** [review]
Excellent job! As WIF is currently only available in private preview, it would be helpful to mention this information somewhere. This way, we can avoid potential issues that may arise when individuals outside the private preview attempt to use it. Moreover, even in public preview, it can be disabled based on the preference of the organization owner.

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go
Comment (jaredfholgate) [reply azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go]: I needed to run these integration tests, I am not sure why they were skipped originally. I will remove the 'skip' altogether and the maintainer can put them back in if needed.

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go
Comment (jaredfholgate) [reply azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go]: Updated.

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go
Comment (jaredfholgate) [reply azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go:148]: I think we can add the draft option later if there is a need for it.

File: azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go
Comment (jaredfholgate) [reply azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go]: Updated.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go
Comment (jaredfholgate) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go:75]: Ok, thanks. We have to keep ServicePrincipal as default for now to avoid a breaking change.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go
Comment (jaredfholgate) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go:262]: Although it is being set here, it is just for the purposes of state management. It is not actually sent to the API.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm_test.go
Comment (jaredfholgate) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm_test.go:185]: If I can run the acceptance tests successfully before this gets merged I will remove it. Otherwise I will raise a separate PR to remove it later.

File: website/docs/r/serviceendpoint_azurerm.html.markdown
Comment (jaredfholgate) [reply website/docs/r/serviceendpoint_azurerm.html.markdown]: This field has to cater for ServicePrincipal too. If they choose the WorkloadIdentityFederation authenticaiton scheme, it will ignore this field and not pass it to the API.

**Comment by jaredfholgate** [comment]
> Excellent job! As WIF is currently only available in private preview, it would be helpful to mention this information somewhere. This way, we can avoid potential issues that may arise when individuals outside the private preview attempt to use it. Moreover, even in public preview, it can be disabled based on the preference of the organization owner.

I added a note to this effect in the docs.

**Review by xuzhang3 — APPROVED** [review]

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go
Comment (jaredfholgate) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go:103]: This is a temporary fix due to a bug in the API. The fix is rolling out soon, but I am unable to run the acceptance test without this line until it does. The value is not important. I have added a comment in the last few minutes to explain this.

File: website/docs/d/serviceendpoint_azurerm.html.markdown
Comment (geekzter) [inline website/docs/d/serviceendpoint_azurerm.html.markdown:1]: @jaredfholgate Can you add the following link: https://aka.ms/azdo-rm-workload-identity. This contains information about the preview and how to get in. Once in public preview, it will redirect to a doc on learn.microsoft.com.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go
Comment (jaredfholgate) [reply azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurerm.go]: I have added a state migration. Let me know if it looks ok, no something I have done before. Thanks.

File: website/docs/d/serviceendpoint_azurerm.html.markdown
Comment (jaredfholgate) [reply website/docs/d/serviceendpoint_azurerm.html.markdown:1]: Done

Comment (jaredfholgate) [reply website/docs/d/serviceendpoint_azurerm.html.markdown:1]: I had added it to the data source instead of the resource by mistake, so moved it now.

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go
Comment (xuzhang3) [inline azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go:187]: the test steps are exactly the same

**Review by xuzhang3 — CHANGES_REQUESTED** [review]

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go
Comment (jaredfholgate) [reply azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go:187]: I copied the existing tests. 

The first step uses the serviceEndpointNameFirst and serviceprincipalidFirst variables.

The second step uses the  serviceEndpointNameSecond and serviceprincipalidSecond variables.

Comment (xuzhang3) [reply azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go:187]: my bad , the duplicate test case is in `TestAccServiceEndpointAzureRm_AutomaticCreateAndUpdate`  line 119~142

Comment (jaredfholgate) [reply azuredevops/internal/acceptancetests/resource_serviceendpoint_azurerm_test.go:187]: I altered that test to add a parameter, but did not change the test logic. If you want me to look into it I can, but it was already like that. Thanks

**Review by xuzhang3 — APPROVED** [review]

**Comment by xuzhang3** [comment]
@jaredfholgate  LGTM

**Pull request merged**
